### ISSUE TITLE
Reflexion::getPropertiesAndValues() might ignore primary key.

### ISSUE THREAD

**Issue by churow** [phpMv/ubiquity#187]
```
$newJob = new Job();
$newJob->setSomething();
DAO::save($newJob);

$oldJob = DAO::getById(Job::class, $id);
$oldJob->setAnotherthing();
DAO::save($oldJob);
```

During the first DAO::save(), Reflexion::getPropertiesAndValues() stores Job's properties into Refelxsion::$classProperties['Job'] except the primary key because it's always null for INSERT when it's auto-incrementing.
During the second DAO::save(), the stored Refelxsion::$classProperties['Job'] is used to return name-property set and the primary key is ignored, which causes number of parameters mismatch while executing query statement.

Because of this issue, I had to reverse the order: DAO::save($oldJob) first, and then DAO::save($newJob).

I think Reflexion::getPropertiesAndValues() needs to check if the call is from INSERT or from UPDATE.

**Comment by jcheron** [comment]
I don't have such a problem, no matter what order the insert and update are in.

The update can hardly have an inconsistent number of parameters: 
it is the same array that is used to build the prepared SQL statement 
https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L323
and then to provide the parameters:
https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L330

The problem must come from elsewhere.
For a better resolution, could you publish the exception encountered, the source code of the model, and its SQL version ?

**Comment by churow** [comment]
Might it be from version difference? I'm currently using Ubiquity 2.3.

These are the abbreviated logs.

First, from the INSERT:
```
{"message":"INSERT INTO `job` (`job_name`,...) VALUES(:job_name,...)","context":{"context":"DAOUpdates","part":"insert","extra":null},"level":200,"level_name":"INFO","channel":"some","datetime":"2021-11-27T12:31:12.223460+09:00","extra":{}}
{"message":"{\"job_name\":\"A_Job_Name\",...}","context":{"context":"DAOUpdates","part":"Key and values","extra":null},"level":200,"level_name":"INFO","channel":"some","datetime":"2021-11-27T12:31:12.801620+09:00","extra":{}}
```
The logged message is from the $keyAndValues and this is also stored into Refelxsion::$classProperties['models\Job'].

https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L210
https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L220

Second, from the UPDATE:
```
{"message":"UPDATE `job` SET `job_name`= :job_name,... WHERE `job_id`= :job_id","context":{"context":"DAOUpdates","part":"update","extra":null},"level":200,"level_name":"INFO","channel":"some","datetime":"2021-11-27T12:36:45.115119+09:00","extra":{}}
{"message":"{\"job_name\":\"Another_Job_Name\",...}","context":{"context":"DAOUpdates","part":"Key and values","extra":null},"level":200,"level_name":"INFO","channel":"some","datetime":"2021-11-27T12:36:45.116101+09:00","extra":{}}
```
The logged message is from the $ColumnskeyAndValues and this is from the previously stored Refelxsion::$classProperties['models\Job']. It lacks the primary key job_id.

https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L321
https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/traits/DAOUpdatesTrait.php#L326

https://github.com/phpMv/ubiquity/blob/100a68a8a77098fb2091fe3d3ce1e292dbe383be/src/Ubiquity/orm/parser/Reflexion.php#L55

The MySQL error was **SQLSTATE[HY093]: Invalid parameter number: number of bound variables does not match number of tokens**.
```
{"message":"SQLSTATE[HY093]: Invalid parameter number: number of bound variables does not match number of tokens","context":{"context":"DAOUpdates","part":"update","extra":null},"level":300,"level_name":"WARNING","channel":"some","datetime":"2021-11-27T12:36:45.116995+09:00","extra":{}}
```

This is the related model:
```
namespace models;
/**
 * @table('job')
*/
class Job{
	/**
	 * @id
	 * @column("name"=>"job_id","nullable"=>false,"dbType"=>"int(10) unsigned")
	 * @validator("id","constraints"=>array("autoinc"=>true))
	**/
	private $job_id;

	/**
	 * @column("name"=>"job_name","nullable"=>false,"dbType"=>"varchar(45)")
	 * @validator("length","constraints"=>array("max"=>45,"notNull"=>true))
	**/
	private $job_name;

	...
}
```

The job table is like this:
```
CREATE TABLE IF NOT EXISTS `job` (
  `job_id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
  `job_name` VARCHAR(45) NOT NULL DEFAULT '',
  ...
  PRIMARY KEY (`job_id`),
  ...
)
ENGINE = InnoDB
AUTO_INCREMENT = 123
DEFAULT CHARACTER SET = utf8mb4
COLLATE = utf8mb4_unicode_ci
```

**Comment by jcheron** [comment]
Yes, it is possible that this bug has been fixed since version 2.3.x
The easiest is to try to switch to a more recent version, at least to know if the bug comes from there.
